File: app/src/main/java/com/hencoder/hencoderpracticedraw2/practice/Practice06LightingColorFilterView.java
package com.hencoder.hencoderpracticedraw2.practice;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ColorFilter;
import android.graphics.LightingColorFilter;
import android.graphics.Paint;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

import com.hencoder.hencoderpracticedraw2.R;

public class Practice06LightingColorFilterView extends View {

    private Paint mPaint;
    private Bitmap mBitmap;

    public Practice06LightingColorFilterView(Context context) {
        this(context, null);
    }

    public Practice06LightingColorFilterView(Context context, @Nullable AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public Practice06LightingColorFilterView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public void init() {
        mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        mBitmap = BitmapFactory.decodeResource(getResources(), R.drawable.batman);
    }


    @SuppressLint("DrawAllocation")
    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);

        // 使用 Paint.setColorFilter() 来设置 LightingColorFilter
        LightingColorFilter colorFilter = new LightingColorFilter(0X00FFFF, 0X000000);
        mPaint.setColorFilter(colorFilter);
        // 第一个 LightingColorFilter：去掉红色部分
        canvas.drawBitmap(mBitmap, 0, 0, mPaint);


        // 第二个 LightingColorFilter：增强绿色部分
        colorFilter = new LightingColorFilter(0XFFFFFF, 0X003000);
        mPaint.setColorFilter(colorFilter);
        canvas.drawBitmap(mBitmap, mBitmap.getWidth() + 100, 0, mPaint);
    }

}

File: app/src/main/java/com/hencoder/hencoderpracticedraw2/practice/Practice06LightingColorFilterView.kt
package com.hencoder.hencoderpracticedraw2.practice

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import com.hencoder.hencoderpracticedraw2.R

/**
 * Created by xh on 2017/10/17.
 */
class Practice06LightingColorFilterView : View {

    private val mPaint = Paint(Paint.ANTI_ALIAS_FLAG)

    private val mBitmap = BitmapFactory.decodeResource(resources, R.drawable.batman)

    private var mLightingColorFilter1 = LightingColorFilter(0X00FFFF, 0)
    private var  mLightingColorFilter2 = LightingColorFilter(0XFFFFFF, 0X003000)

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)
    constructor(context: Context?, attrs: AttributeSet?, defStyleAttr: Int) : super(context, attrs, defStyleAttr)

    override fun onDraw(canvas: Canvas?) {
        super.onDraw(canvas)

        mPaint.colorFilter = mLightingColorFilter1
        canvas!!.drawBitmap(mBitmap, 0f, 0f, mPaint)
        mPaint.colorFilter = mLightingColorFilter2
        canvas.drawBitmap(mBitmap, mBitmap.width + 100f, 0f, mPaint)

    }

}